promql/engine.go

```diff
@@ -697,117 +655,355 @@ func (ev *evaluator) eval(expr Expr) Value {
+					ss = Series{
+						Metric: sample.Metric,
+						Points: getPointSlice(numSteps),
+					}
+					Seriess[h] = ss
+				}
+				sample.Point.T = ts
+				ss.Points = append(ss.Points, sample.Point)
+				Seriess[h] = ss
+			}
+		}
+		// Reuse the original point slices.
+		for _, m := range origMatrixes {
+			for _, s := range m {
+				putPointSlice(s.Points)
+			}
+		}
+		// Assemble the output matrix.
+		mat := Matrix{}
+		for _, ss := range Seriess {
+			mat = append(mat, ss)
+		}
+		return mat
+	}
 
 	switch e := expr.(type) {
 	case *AggregateExpr:
-		Vector := ev.evalVector(e.Expr)
-		return ev.aggregation(e.Op, e.Grouping, e.Without, e.Param, Vector)
+		if s, ok := e.Param.(*StringLiteral); ok {
+			return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+				return ev.aggregation(e.Op, e.Grouping, e.Without, s.Val, v[0].(Vector), enh)
+			}, e.Expr)
+		}
+		return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+			var param float64
+			if e.Param != nil {
+				param = v[0].(Vector)[0].V
+			}
+			return ev.aggregation(e.Op, e.Grouping, e.Without, param, v[1].(Vector), enh)
+		}, e.Param, e.Expr)
 
-	case *BinaryExpr:
-		lhs := ev.evalOneOf(e.LHS, ValueTypeScalar, ValueTypeVector)
-		rhs := ev.evalOneOf(e.RHS, ValueTypeScalar, ValueTypeVector)
+	case *Call:
+		if e.Func.Name == "timestamp" {
+			// Matrix evaluation always returns the evaluation time,
+			// so this function needs special handling when given
+			// a vector selector.
+			vs, ok := e.Args[0].(*VectorSelector)
+			if ok {
+				return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+					return e.Func.Call([]Value{ev.vectorSelector(vs, enh.ts)}, e.Args, enh)
+				})
+			}
+		}
+		// Check if the function has a matrix argument.
+		for matrixArgIndex, a := range e.Args {
+			sel, ok := a.(*MatrixSelector)
+			if ok {
+				inArgs := make([]Value, len(e.Args))
+				// Evaluate any non-matrix arguments.
+				otherArgs := make([]Matrix, len(e.Args))
+				otherArgsIn := make([]Vector, len(e.Args))
+				for i, e := range e.Args {
+					if i != matrixArgIndex {
+						otherArgs[i] = ev.eval(e).(Matrix)
+						otherArgsIn[i] = Vector{Sample{}}
+						inArgs[i] = otherArgsIn[i]
+					}
+				}
 
-		switch lt, rt := lhs.Type(), rhs.Type(); {
-		case lt == ValueTypeScalar && rt == ValueTypeScalar:
-			return Scalar{
-				V: scalarBinop(e.Op, lhs.(Scalar).V, rhs.(Scalar).V),
-				T: ev.Timestamp,
+				mat := make(Matrix, 0, len(sel.series)) // Output matrix.
+				offset := durationMilliseconds(sel.Offset)
+				// Reuse objects across steps to save memory allocations.
+				points := getPointSlice(16)
+				inMatrix := make(Matrix, 1)
+				inArgs[matrixArgIndex] = inMatrix
+				enh := &evalNodeHelper{out: make(Vector, 0, 1)}
+				// Process all the calls for one time series at a time.
+				var it *storage.BufferedSeriesIterator
+				for i, s := range sel.series {
+					if it == nil {
+						it = storage.NewBuffer(s.Iterator(), durationMilliseconds(sel.Range))
```

Comment:
At this point it looks like it may be useful to have a

    storage.NewBuffer(durationMilliseconds(sel.Range))

factory function that initializes the underlying iterator with a dummy. Then you wouldn't need to repeat this if-else all over the place, you could just call `Reset()` directly.

Reply:
That's presuming there's only one iterator at a time. This is something that needs to be figured out generally, as I want to do this in several other places to reduce allocations.

Reply:
As the code is now you are actually iterating over each series in turn, so you do have one iterator at a time.

But I don't think it makes any difference whether you're scanning one series or more, as long as you're using the same range for all. Am I missing something?

It's just a way of allocating the buffer ahead of time (which will be unnecessary when you have zero series) so you don't always have to check for `nil`.

Reply:
Like I said, I don't see how this has anything to do with one vs. more iterators. (Although I might be missing something.)

All I was trying to say was that it would be helpful to have a [zero value](https://golang.org/ref/spec#The_zero_value) for `storage.BufferedSeriesIterator`, so you don't always have to check for `nil` (which happens repeatedly throughout the code).

Reply:
I'll keep it in mind for future work (which will be doing more of this), but I don't want to expand the scope of this PR further.